tests/src/test/java/io/streamnative/pulsar/handlers/kop/MetricsProviderTest.java

```diff
@@ -230,36 +234,65 @@ public void testMetricsProvider() throws Exception {
-        byte[] data = mockZooKeeper.getData(conf.getGroupIdZooKeeperPath() + "/" + children.get(0), false, null);
-        Assert.assertEquals(new String(data, StandardCharsets.UTF_8), group1);
+        tryConsume(topic, clientId, group1, SafeRunnable.safeRun(() -> {
+            try {
+                List<String> children = mockZooKeeper.getChildren(conf.getGroupIdZooKeeperPath(), false);
+                Assert.assertEquals(children.size(), 1);
+                Assert.assertEquals(children.get(0), "127.0.0.1-" + clientId);
+                byte[] data = mockZooKeeper
+                        .getData(conf.getGroupIdZooKeeperPath() + "/" + children.get(0), false, null);
+                Assert.assertEquals(new String(data, StandardCharsets.UTF_8), group1);
+            } catch (Exception ex) {
+                fail("Should not have exception." + ex.getMessage());
+            }
+        }));
+
+        Awaitility.await().untilAsserted(() -> {
+            List<String> children1 = mockZooKeeper.getChildren(conf.getGroupIdZooKeeperPath(), false);
+            Assert.assertEquals(children1.size(), 0);
+        });
 
         // Create a consumer with the same hostname and client id, the existed z-node will be updated
-        tryConsume(topic, clientId, group2);
-        children = mockZooKeeper.getChildren(conf.getGroupIdZooKeeperPath(), false);
-        Assert.assertEquals(children.size(), 1);
-        Assert.assertEquals(children.get(0), "127.0.0.1-" + clientId);
-        data = mockZooKeeper.getData(conf.getGroupIdZooKeeperPath() + "/" + children.get(0), false, null);
-        Assert.assertEquals(new String(data, StandardCharsets.UTF_8), group2);
+        tryConsume(topic, clientId, group2, SafeRunnable.safeRun(() -> {
+            try {
+                List<String> children = mockZooKeeper.getChildren(conf.getGroupIdZooKeeperPath(), false);
+                Assert.assertEquals(children.size(), 1);
+                Assert.assertEquals(children.get(0), "127.0.0.1-" + clientId);
+                byte[] data = mockZooKeeper
+                        .getData(conf.getGroupIdZooKeeperPath() + "/" + children.get(0), false, null);
+                Assert.assertEquals(new String(data, StandardCharsets.UTF_8), group2);
+            } catch (Exception ex) {
+                fail("Should not have exception." + ex.getMessage());
+            }
+        }));
+
+        Awaitility.await().untilAsserted(() -> {
+            List<String> children1 = mockZooKeeper.getChildren(conf.getGroupIdZooKeeperPath(), false);
+            Assert.assertEquals(children1.size(), 0);
+        });
     }
 
-    private void tryConsume(final String topic, final String clientId, final String groupId) {
+    private void tryConsume(final String topic,
+                            final String clientId,
+                            final String groupId,
+                            SafeRunnable runBeforeClose) {
```

Comment:
We should not use `SafeRunnable` because it swallows all exceptions. In addition, since checked in a `Runnable` must be caught, using a `Runnable` here is too complicated because we need a try-catch block.

I think 

```java
    public void testUpdateGroupId() throws Exception {
        // ...
        tryConsume(topic, clientId, group1);
        List<String> children = mockZooKeeper.getChildren(conf.getGroupIdZooKeeperPath(), false);
        Assert.assertEquals(children.size(), 1);
        Assert.assertEquals(children.get(0), "127.0.0.1-" + clientId);
        byte[] data = mockZooKeeper
                .getData(conf.getGroupIdZooKeeperPath() + "/" + children.get(0), false, null);
        Assert.assertEquals(new String(data, StandardCharsets.UTF_8), group1);
```

looks more clear than

```java
    public void testUpdateGroupId() {
        // ...
        tryConsume(topic, clientId, group1, () -> {
            try {
                List<String> children = mockZooKeeper.getChildren(conf.getGroupIdZooKeeperPath(), false);
                Assert.assertEquals(children.size(), 1);
                Assert.assertEquals(children.get(0), "127.0.0.1-" + clientId);
                byte[] data = mockZooKeeper
                        .getData(conf.getGroupIdZooKeeperPath() + "/" + children.get(0), false, null);
                Assert.assertEquals(new String(data, StandardCharsets.UTF_8), group1);
            } catch (Exception ex) {
```

Reply:
We need to check if the path exists before the client close. Can you explain it in detail?

Reply:
Sorry I missed this point.